build/lib/src/generate/expected_outputs.dart

```diff
@@ -52,31 +55,97 @@ extension on AssetId {
 }
 
 abstract class _ParsedBuildOutputs {
-  static final RegExp _captureGroup = RegExp('{{(\\w*)}}');
-
   _ParsedBuildOutputs();
 
   factory _ParsedBuildOutputs.parse(
       Builder builder, String input, List<String> outputs) {
-    final matches = _captureGroup.allMatches(input).toList();
-
-    if (matches.isEmpty) {
-      // The input does not contain a capture group, so we should simply match
-      // all assets whose paths ends with the desired input.
-      // Also, make sure that no outputs use capture groups.
-      for (final output in outputs) {
-        if (_captureGroup.hasMatch(output)) {
-          throw ArgumentError(
-            'The builder `$builder` declares an output "$output" using a '
-            'capture group. As its input "$input" does not use a capture '
-            'group, this is forbidden.',
-          );
-        }
+    final matches = _captureGroupRegexp.allMatches(input).toList();
+    if (matches.isNotEmpty) {
+      return _CapturingBuildOutputs.parse(builder, input, outputs, matches);
+    }
+
+    // Make sure that no outputs use capture groups, if they aren't used in
+    // inputs.
+    for (final output in outputs) {
+      if (_captureGroupRegexp.hasMatch(output)) {
+        throw ArgumentError(
+          'The builder `$builder` declares an output "$output" using a '
+          'capture group. As its input "$input" does not use a capture '
+          'group, this is forbidden.',
+        );
       }
+    }
 
+    if (input.startsWith('^')) {
+      return _FullMatchBuildOutputs(input.substring(1), outputs);
+    } else {
       return _SuffixBuildOutputs(input, outputs);
     }
+  }
+
+  bool hasAnyOutputFor(AssetId input);
+  Iterable<AssetId> matchingOutputsFor(AssetId input);
+}
+
+/// A simple build input/output set that matches an entire path and doesn't use
+/// capture groups.
+class _FullMatchBuildOutputs extends _ParsedBuildOutputs {
+  final String inputExtension;
+  final List<String> outputExtensions;
+
+  _FullMatchBuildOutputs(this.inputExtension, this.outputExtensions);
+
+  @override
+  bool hasAnyOutputFor(AssetId input) => input.path == inputExtension;
+
+  @override
+  Iterable<AssetId> matchingOutputsFor(AssetId input) {
+    if (!hasAnyOutputFor(input)) return const Iterable.empty();
+
+    // If we expect an output, the asset's path ends with the input extension.
+    // Expected outputs just replace the matched suffix in the path.
+    return outputExtensions
+        .map((extension) => AssetId(input.package, extension));
+  }
+}
+
+/// A simple build input/output set which matches file suffixes and doesn't use
```

Comment:
This was not actually edited... the diff just got weird because of how I moved some of the other code around 🤷‍♂️